python/cudf/cudf/core/frame.py

```diff
@@ -97,6 +97,274 @@ def _as_column(self):
 
         return self._data[None].copy(deep=False)
 
+    def _scatter(self, key, value):
+        result = self._from_table(libcudfxx.copying.scatter(value, key, self))
+
+        result._copy_categories(self)
+        return result
+
+    def _empty_like(self):
+        result = self._from_table(libcudfxx.copying.table_empty_like(self))
+
+        result._copy_categories(self)
+        return result
+
+    def _slice(self, arg):
+        """
+       _slice : slice the frame as per the arg
+
+       Parameters
+       ----------
+       arg : should always be of type slice and doesn't handle step
+
+       """
+        num_rows = len(self)
+        if num_rows == 0:
+            return self
+        start, stop, stride = arg.indices(num_rows)
+
+        if stride is not None and stride != 1:
+            raise ValueError(
+                """Step size is not supported other than None and 1"""
+            )
+
+        if start < 0:
+            start = start + num_rows
+        if stop < 0:
+            stop = stop + num_rows
+
+        if start > stop:
+            return self._empty_like()
+        else:
+            start = len(self) if start > num_rows else start
+            stop = len(self) if stop > num_rows else stop
+
+            result = self._from_table(
+                libcudfxx.copying.table_slice(self, [start, stop])[0]
+            )
+
+            result._copy_categories(self)
+            return result
+
+    def _normalize_scalars(self, other):
+        """
+        Try to normalizes scalar values as per self dtype
+        """
+        if (
+            other is not None
+            and (isinstance(other, float) and not np.isnan(other))
+        ) and (self.dtype.type(other) != other):
+            raise TypeError(
+                "Cannot safely cast non-equivalent {} to {}".format(
+                    type(other).__name__, self.dtype.name
+                )
+            )
+
+        return (
+            self.dtype.type(other)
+            if (
+                other is not None
+                and (isinstance(other, float) and not np.isnan(other))
+            )
+            else other
+        )
+
+    def _normalize_columns_and_scalars_type(self, other):
+        """
+        Try to normalize the other's dtypes as per self.
+
+        Parameters
+        ----------
+
+        self : Can be a DataFrame or Series or Index
+        other : Can be a DataFrame, Series, Index, Array
+                like object or a scalar value
+
+                if self is DataFrame, other can be only a
+                scalar or array like with size of number of columns
+                in DataFrame or a DataFrame with same dimenstion
+
+                if self is Series, other can be only a scalar or
+                a series like with same length as self
```

Comment:
docstring formatting: after first line following lines are indented four spaces, not aligned with first line